<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head><title>R: Phylogeny Summary Statistics Demonstrated</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>
<h1> Phylogeny Summary Statistics Demonstrated
<img class="toplogo" src="../../../doc/html/Rlogo.svg" alt="[R logo]" />
</h1>
<hr/>
<div style="text-align: center;">
<a href="../../../doc/html/packages.html"><img class="arrow" src="../../../doc/html/left.jpg" alt="[Up]" /></a>
<a href="../../../doc/html/index.html"><img class="arrow" src="../../../doc/html/up.jpg" alt="[Top]" /></a>
</div><h2>Documentation for package &lsquo;Cer2016&rsquo; version 0.1.0</h2>

<ul><li><a href="../DESCRIPTION">DESCRIPTION file</a>.</li>
</ul>

<h2>Help Pages</h2>


<table width="100%">
<tr><td style="width: 25%;"><a href="add_alignments.html">add_alignments</a></td>
<td>Add an alignment to a file</td></tr>
<tr><td style="width: 25%;"><a href="add_outgroup_to_phylogeny.html">add_outgroup_to_phylogeny</a></td>
<td>Adds an outgroup to a phylogeny</td></tr>
<tr><td style="width: 25%;"><a href="add_pbd_output.html">add_pbd_output</a></td>
<td>Adds a pbd_sim result to a file</td></tr>
<tr><td style="width: 25%;"><a href="add_posteriors.html">add_posteriors</a></td>
<td>Add BEAST2 posteriors to a file</td></tr>
<tr><td style="width: 25%;"><a href="add_species_trees_with_outgroup.html">add_species_trees_with_outgroup</a></td>
<td>Add a species tree with outgroup to a file</td></tr>
<tr><td style="width: 25%;"><a href="branch_sorter.html">branch_sorter</a></td>
<td>Sorts branches for the NRBS analysis</td></tr>
<tr><td style="width: 25%;"><a href="branch_sorter_from_file.html">branch_sorter_from_file</a></td>
<td>sorts branches for the NRBS analysis</td></tr>
<tr><td style="width: 25%;"><a href="check_progress.html">check_progress</a></td>
<td>Function that shows how much parameter files are present, and how far the pipeline has gotten</td></tr>
<tr><td style="width: 25%;"><a href="collect_files_gammas.html">collect_files_gammas</a></td>
<td>Collects the gamma statistics of all phylogenies belonging to a multiple parameter file in the melted/uncast/long form</td></tr>
<tr><td style="width: 25%;"><a href="collect_file_gammas.html">collect_file_gammas</a></td>
<td>Collects the gamm statistics of all phylogenies belonging to a parameter file in the melted/uncast/long form</td></tr>
<tr><td style="width: 25%;"><a href="collect_gamma_statistics.html">collect_gamma_statistics</a></td>
<td>Collects the gamma statistics of all phylogenies in the melted/uncast/long form</td></tr>
<tr><td style="width: 25%;"><a href="collect_parameters.html">collect_parameters</a></td>
<td>Creates tidy data of all parameter files</td></tr>
<tr><td style="width: 25%;"><a href="collect_posterior_filenames.html">collect_posterior_filenames</a></td>
<td>Collect posterior filenames</td></tr>
<tr><td style="width: 25%;"><a href="collect_posterior_gammas.html">collect_posterior_gammas</a></td>
<td>Collect the gamma statistics of the BEAST2 posteriors</td></tr>
<tr><td style="width: 25%;"><a href="collect_posterior_nltts.html">collect_posterior_nltts</a></td>
<td>Collect the nLTT values of the BEAST2 posteriors</td></tr>
<tr><td style="width: 25%;"><a href="collect_species_tree_gammas.html">collect_species_tree_gammas</a></td>
<td>Collect the gamma statistics of the species trees with outgroup</td></tr>
<tr><td style="width: 25%;"><a href="collect_species_tree_nltts.html">collect_species_tree_nltts</a></td>
<td>Collect the nLTT values of the species trees with outgroup</td></tr>
<tr><td style="width: 25%;"><a href="convert_alignment_to_beast_input_file.html">convert_alignment_to_beast_input_file</a></td>
<td>Convert an alignment and parameters to a BEAST XML input file</td></tr>
<tr><td style="width: 25%;"><a href="convert_alignment_to_beast_posterior.html">convert_alignment_to_beast_posterior</a></td>
<td>Convert an alignment to a BEAST2 XML input file</td></tr>
<tr><td style="width: 25%;"><a href="convert_alignment_to_fasta.html">convert_alignment_to_fasta</a></td>
<td>Convert an alignment (of type DNAbin) to a FASTA file</td></tr>
<tr><td style="width: 25%;"><a href="convert_phylogeny_to_alignment.html">convert_phylogeny_to_alignment</a></td>
<td>Converts a phylogeny to a random DNA alignment</td></tr>
<tr><td style="width: 25%;"><a href="extract_erg.html">extract_erg</a></td>
<td>Extract the ERG ('Extinction Rate of a Good species') parameter value from a file</td></tr>
<tr><td style="width: 25%;"><a href="extract_eri.html">extract_eri</a></td>
<td>Extract the ERI ('Extinction Rate of an Incipient species') parameter value from a file</td></tr>
<tr><td style="width: 25%;"><a href="extract_scr.html">extract_scr</a></td>
<td>Extract the SCR ('Speciation Completion Rate') parameter value from a file</td></tr>
<tr><td style="width: 25%;"><a href="extract_sirg.html">extract_sirg</a></td>
<td>Extract the SIRG ('Speciation Initial Rate of a Good species') parameter value from a file</td></tr>
<tr><td style="width: 25%;"><a href="extract_siri.html">extract_siri</a></td>
<td>Extract the SIRI ('Speciation Initial Rate of an Incipient species') parameter value from a file</td></tr>
<tr><td style="width: 25%;"><a href="find_beast_bin_path.html">find_beast_bin_path</a></td>
<td>Find the path of the BEAST2 binary file</td></tr>
<tr><td style="width: 25%;"><a href="find_beast_jar_path.html">find_beast_jar_path</a></td>
<td>Find the path of the BEAST2 jar file</td></tr>
<tr><td style="width: 25%;"><a href="find_beast_posterior_test_filename.html">find_beast_posterior_test_filename</a></td>
<td>Find the path of known-to-be-valid BEAST2 posterior file</td></tr>
<tr><td style="width: 25%;"><a href="find_path.html">find_path</a></td>
<td>Finds the full pathof a file</td></tr>
<tr><td style="width: 25%;"><a href="get_phylogeny_crown_age.html">get_phylogeny_crown_age</a></td>
<td>Obtain the crown age of a phylony</td></tr>
<tr><td style="width: 25%;"><a href="is_alignment.html">is_alignment</a></td>
<td>Determines if the input is an alignment of type DNAbin</td></tr>
<tr><td style="width: 25%;"><a href="is_beast_posterior.html">is_beast_posterior</a></td>
<td>Determines if the input is a BEAST2 posterior</td></tr>
<tr><td style="width: 25%;"><a href="is_pbd_sim_output.html">is_pbd_sim_output</a></td>
<td>Does not use pbd_sim()$stree, but generates these like PBD does</td></tr>
<tr><td style="width: 25%;"><a href="is_phylogeny.html">is_phylogeny</a></td>
<td>Checks if the input is a phylogeny</td></tr>
<tr><td style="width: 25%;"><a href="is_valid_file.html">is_valid_file</a></td>
<td>Checks if a file is a valid parameter file</td></tr>
<tr><td style="width: 25%;"><a href="is_whole_number.html">is_whole_number</a></td>
<td>Checks if x is a single, whole-number variable</td></tr>
<tr><td style="width: 25%;"><a href="load_parameters_from_file.html">load_parameters_from_file</a></td>
<td>Load parameters from file</td></tr>
<tr><td style="width: 25%;"><a href="normalized_rooted_branch_score.html">normalized_rooted_branch_score</a></td>
<td>calculates the normalized rooted branch score</td></tr>
<tr><td style="width: 25%;"><a href="plot_alignments.html">plot_alignments</a></td>
<td>Plot alignments</td></tr>
<tr><td style="width: 25%;"><a href="plot_extant_incipient_tree.html">plot_extant_incipient_tree</a></td>
<td>Plot the extant inicipient species tree</td></tr>
<tr><td style="width: 25%;"><a href="plot_posterior_nltts.html">plot_posterior_nltts</a></td>
<td>Plot multiple BEAST2 posteriors</td></tr>
<tr><td style="width: 25%;"><a href="plot_posterior_samples.html">plot_posterior_samples</a></td>
<td>Plot a random posterior phylogeny</td></tr>
<tr><td style="width: 25%;"><a href="plot_posterior_sample_nltts.html">plot_posterior_sample_nltts</a></td>
<td>Plot the nLTTs of the posterior</td></tr>
<tr><td style="width: 25%;"><a href="plot_species_tree_with_outgroup.html">plot_species_tree_with_outgroup</a></td>
<td>Plot</td></tr>
<tr><td style="width: 25%;"><a href="plot_species_tree_with_outgroup_nltt.html">plot_species_tree_with_outgroup_nltt</a></td>
<td>Plot species_tree_with_outgroup_nltt</td></tr>
<tr><td style="width: 25%;"><a href="read_file.html">read_file</a></td>
<td>Read a parameter (.RDa) file</td></tr>
<tr><td style="width: 25%;"><a href="sample_species_trees_from_pbd_sim_output.html">sample_species_trees_from_pbd_sim_output</a></td>
<td>Samples species trees from PBD::pbd_sim its output It does not use pbd_sim()$stree, but generates these like PBD does.</td></tr>
<tr><td style="width: 25%;"><a href="save_parameters_to_file.html">save_parameters_to_file</a></td>
<td>Creates a valid parameter file</td></tr>
<tr><td style="width: 25%;"><a href="show_parameter_files.html">show_parameter_files</a></td>
<td>Creates a nice knitr::table to show one or more parameter files</td></tr>
</table>
</body></html>
